File: cpp/134.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <deque>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int canCompleteCircuit(vector<int> &gas, vector<int> &cost) {
    if(accumulate(gas.begin(), gas.end(), 0) < accumulate(cost.begin(), cost.end(), 0))
      return -1;
    int start = 0, vol = 0;
    for(uint i=0; i<gas.size(); i++){
      vol += gas[i] - cost[i];
      if(vol < 0){
        start = i+1;
        vol = 0;
      }
    }
    return start;
  }
};

int main(){
  auto sol = Solution();
}

File: python/134.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        n = len(gas)
        gas = gas + gas
        cost = cost + cost
        start = 0
        cur = 0
        for i in xrange(n+n):
            cur += gas[i] - cost[i]
            if cur < 0:
                cur = 0
                start = i+1
            else:
                if i+1 - start == n:
                    return start
        return -1
